### PR TITLE
[System Probe] Add logging for offset guessing

### PR DESCRIPTION
**Pull request by stb2132**
### What does this PR do?

Adds offset guessing logging for setting up the system probe. 

### Motivation

A current support request and future support investigations.

### CODE REVIEW

File: pkg/ebpf/c/tracer-ebpf.c [hunk 1]
```diff
@@ -796,8 +796,11 @@ int kprobe__tcp_get_info(struct pt_regs* ctx) {
         return 0;
     }
 
-    guess_offsets(status, sk);
+    if (status->tcp_info_kprobe_status== 0) {
```
Comment (sunhay) [inline pkg/ebpf/c/tracer-ebpf.c:799]: formatting nit: `status->tcp_info_kprobe_status == 0`

File: pkg/ebpf/c/tracer-ebpf.c [hunk 2]
```diff
@@ -796,8 +796,11 @@ int kprobe__tcp_get_info(struct pt_regs* ctx) {
         return 0;
     }
 
-    guess_offsets(status, sk);
+    if (status->tcp_info_kprobe_status == 0) {
+         status->tcp_info_kprobe_status = 1;
+    }
```
[the remark this answers is not in the document]
Comment (stb2132) [reply pkg/ebpf/c/tracer-ebpf.c:801]: That's fair - we can just set it to 1 every time this is called. For some reason, I thought that we should check if we haven't already set it.

File: pkg/ebpf/offsetguess.go [hunk 3]
```diff
@@ -412,6 +431,7 @@ func guessOffsets(m *elf.Module, cfg *Config) error {
 	}
 
 	for status.state != stateReady {
+		//2020-04-16 09:invalid guessing state while guessing source address, got checking expected checked
```
[the remark this answers is not in the document]
Comment (stb2132) [reply pkg/ebpf/offsetguess.go:434]: I did not, thank you for pointing this out.

**Review by leeavital — APPROVED** [review]

File: go.sum [hunk 4]
```diff
@@ -748,6 +748,7 @@ github.com/jsonnet-bundler/jsonnet-bundler v0.1.1-0.20190930114713-10e24cb86976/
 github.com/jsonnet-bundler/jsonnet-bundler v0.2.0/go.mod h1:/by7P/OoohkI3q4CgSFqcoFsVY+IaNbzOVDknEsKDeU=
 github.com/jstemmer/go-junit-report v0.0.0-20190106144839-af01ea7f8024/go.mod h1:6v2b51hI/fHJwM22ozAgKL4VKDeJcHhJFhtBdhmNjmU=
 github.com/jstemmer/go-junit-report v0.9.1/go.mod h1:Brl9GWCQeLvo8nXZwPNNblvFj/XSXhF0NWZEnDohbsk=
+github.com/jteeuwen/go-bindata v0.0.0-20151023091102-a0ff2567cfb7 h1:fNOeTlrOg7VMXyGgvnHoqw6owD7hd3mKSlivApPJMzI=
```
[the remark this answers is not in the document]
Comment (stb2132) [reply go.sum:751]: I'm not 100%, but I've done a bit of research into this and I think it's because a new module version of `go-bindata` was used to build the system probe on the agent dev box. If I remove this line and build again, then this line is always added to `go.sum`. 

Looking at the go modules documentation, the "go.sum contains the expected cryptographic checksums of the content of specific module versions," so I don't think it will hurt anything to include this value. Documentation here:

- https://github.com/golang/go/wiki/Modules#should-i-commit-my-gosum-file-as-well-as-my-gomod-file

Worst case, it is just removed at a later point in time.

[the remark this answers is not in the document]
Comment (leeavital) [reply go.sum:756]: I'm pretty sure this is because `inv -e system-probe.build` downloads its own version of go-bindata. 

This should be fixed by https://github.com/DataDog/datadog-agent/pull/5792   .

File: pkg/ebpf/c/tracer-ebpf.h [hunk 5]
```diff
@@ -102,6 +102,7 @@ typedef struct {
 
 typedef struct {
     __u64 state;
+    __u64 tcp_info_kprobe_status;
```
Comment (leeavital) [inline pkg/ebpf/c/tracer-ebpf.h:105]: Let's add a comment here explaining 0=not triggered, 1=triggered

File: pkg/ebpf/offsetguess.go [hunk 6]
```diff
@@ -359,6 +375,7 @@ func setReadyState(m *elf.Module, mp *elf.Map, status *tracerStatus) error {
 // guess the next field.
 func guessOffsets(m *elf.Module, cfg *Config) error {
 	mp := m.Map(string(tracerStatusMap))
+	threshold = cfg.OffsetGuessThreshold
```
Comment (leeavital) [inline pkg/ebpf/offsetguess.go:378]: Let's pass the threshold into `checkAndUpdateCurrentOffset` instead of storing threshold as a global variable.

File: pkg/ebpf/offsetguess.go [hunk 7]
```diff
@@ -539,3 +557,11 @@ func tcpGetInfo(conn net.Conn) (*syscall.TCPInfo, error) {
...
 	return &tcpInfo, nil
 }
+
+func logSuccessfulGuess(guess C.__u64, offset C.__u64) {
+	if guess == guessDaddrIPv6 {
+		log.Debugf("Successfully guessed %v with offset of %d bytes from the previous field. Guessing complete.", whatString[guess], offset)
+	} else {
+		log.Debugf("Successfully guessed %v with offset of %d bytes from the previous field. Now guessing %v", whatString[guess], offset, whatString[guess+1])
+	}
+}
```
Comment (leeavital) [inline pkg/ebpf/offsetguess.go:567]: As discussed over DM, this is a little bit brittle because it hardcodes the order of offsetguessing to have guessDaddrIPv6 last. It would be cleaner to have two logs -- one for starting offset guessing, one for finishing.

File: pkg/ebpf/offsetguess.go [hunk 8]
```diff
@@ -80,9 +76,20 @@ var whatString = map[C.__u64]string{
 	guessSport:     "source port",
 	guessDport:     "destination port",
 	guessNetns:     "network namespace",
+	guessRTT:       "Round Trip Time",
 	guessDaddrIPv6: "destination address IPv6",
 }
 
+const (
+	tcpInfoKProbeNotCalled C.__u64 = 0
```
Comment (truthbk) [inline pkg/ebpf/offsetguess.go:84]: For consistency shouldn't both (or neither) be `C.__u64` - on assignment to the `tcpKprobeCalledString` they might be transparently casted.

File: pkg/process/config/yaml_config.go [hunk 9]
```diff
@@ -80,6 +80,16 @@ func (a *AgentConfig) loadSysProbeYamlConfig(path string) error {
 		a.ConntrackRateLimit = config.Datadog.GetInt(key(spNS, "conntrack_rate_limit"))
 	}
 
+	// When reading kernel structs at different offsets, don't go over the threshold
+	// This defaults to 400 and has a max of 3000. These are arbitrary choices to avoid infinite loops.
+	if th := config.Datadog.GetInt(key(spNS, "offset_guess_threshold")); th > 0 {
+		if th < maxOffsetThreshold {
+			a.OffsetGuessThreshold = uint64(th)
+		} else {
+			log.Warn("Overriding the configured offset_guess_threshold because it exceeds maximum of 3000")
```
Comment (truthbk) [inline pkg/process/config/yaml_config.go:89]: If the setting is exceeded, should you set `a.OffsetGuessThreshold` to the default(400) or to the max(3000)?
Also, I feel like the word `overriding` here is a bit misleading, I'd maybe reword and print the actual value used for the setting.

File: pkg/process/config/config.go [hunk 10]
```diff
@@ -207,6 +209,7 @@ func NewDefaultAgentConfig(canAccessContainers bool) *AgentConfig {
 		ClosedChannelSize:     500,
 		ConntrackMaxStateSize: defaultMaxTrackedConnections * 2,
 		ConntrackRateLimit:    500,
+		OffsetGuessThreshold:  400,
```
Comment (truthbk) [inline pkg/process/config/config.go:212]: I know this is the case for other settings that were already present as well - so I'm not entirely sure this is the right PR to address this, but probably something you need to keep in mind for the future - we're repeating this `400` number all over the place, it should be a `const` in one place, and should be referenced everywhere else (kind of like what you've done with the `maxOffsetThreshold`). If you can fix it now cleanly, and set an example on how this should be done for all similar setting in the sysprobe, it would be great.

File: pkg/ebpf/c/tracer-ebpf.c [hunk 11]
```diff
@@ -796,8 +796,8 @@ int kprobe__tcp_get_info(struct pt_regs* ctx) {
         return 0;
     }
 
+    status->tcp_info_kprobe_status = 1;
```
Comment (truthbk) [inline pkg/ebpf/c/tracer-ebpf.c:799]: Just double-checking, is default triggered (1)?

**Review by truthbk — COMMENTED** [review]
Mostly looks good to me (from the Agent core side) given @leeavital feedback. Just added some minor comments.

File: pkg/process/config/yaml_config.go (hunk 9 above)
Comment (stb2132) [reply pkg/process/config/yaml_config.go:89]: Hey @truthbk, in other situations in which we have set a value too high, we use the default, so I was maintaining that behavior for consistency's sake.

File: pkg/ebpf/offsetguess.go (hunk 8 above)
Comment (stb2132) [reply pkg/ebpf/offsetguess.go:84]: So I kept it this way for consistency with the rest of the file.

File: pkg/ebpf/c/tracer-ebpf.c (hunk 11 above)
Comment (stb2132) [reply pkg/ebpf/c/tracer-ebpf.c:799]: What do you mean by default here? This should be 0 initially and only set to 1 when this function is invoked. 

I previously had it in an if statement, but basically this is just a way to determine (we have a support case where we need to check this currently) that this kprobe runs at all.

File: pkg/process/config/config.go (hunk 10 above)
Comment (stb2132) [reply pkg/process/config/config.go:212]: Hey, apologies, but I don't think I can address this in this PR. I would need to export it to access it in another struct in the ebpf package, which I don't want to arbitrarily do. 

Since this is largely for support, I don't want to add the extra complexity and potentially miss getting this into 7.21 (if that's okay).

**Review by truthbk — APPROVED** [review]

**Review by derekwbrown — APPROVED** [review]
